crates/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -124,29 +132,43 @@ fn run_test<B: Backend>(
     backend: &B,
     test_name: &str,
     main: FuncId,
+    should_fail: bool,
     context: &Context,
     show_output: bool,
     config: &CompileOptions,
 ) -> Result<(), CliError<B>> {
-    let mut program = compile_no_check(context, config, main).map_err(|err| {
-        noirc_errors::reporter::report_all(&context.file_manager, &[err], config.deny_warnings);
-        CliError::Generic(format!("Test '{test_name}' failed to compile"))
-    })?;
-
-    // Note: We could perform this test using the unoptimized ACIR as generated by `compile_no_check`.
-    program.circuit = optimize_circuit(backend, program.circuit).unwrap().0;
-
-    // Run the backend to ensure the PWG evaluates functions like std::hash::pedersen,
-    // otherwise constraints involving these expressions will not error.
-    match execute_circuit(backend, program.circuit, WitnessMap::new(), show_output) {
-        Ok(_) => Ok(()),
-        Err(error) => {
-            let writer = StandardStream::stderr(ColorChoice::Always);
-            let mut writer = writer.lock();
-            writer.set_color(ColorSpec::new().set_fg(Some(Color::Red))).ok();
-            writeln!(writer, "failed").ok();
-            writer.reset().ok();
-            Err(error.into())
+    if !should_fail {
+        let mut program = compile_no_check(context, config, main).map_err(|err| {
+            noirc_errors::reporter::report_all(&context.file_manager, &[err], config.deny_warnings);
+            CliError::Generic(format!("Test '{test_name}' failed to compile"))
+        })?;
+
+        // Note: We could perform this test using the unoptimized ACIR as generated by `compile_no_check`.
+        program.circuit = optimize_circuit(backend, program.circuit).unwrap().0;
+
+        // Run the backend to ensure the PWG evaluates functions like std::hash::pedersen,
+        // otherwise constraints involving these expressions will not error.
+        match execute_circuit(backend, program.circuit, WitnessMap::new(), show_output) {
+            Ok(_) => Ok(()),
+            Err(error) => {
+                let writer = StandardStream::stderr(ColorChoice::Always);
+                let mut writer = writer.lock();
+                writer.set_color(ColorSpec::new().set_fg(Some(Color::Red))).ok();
+                writeln!(writer, "failed").ok();
+                writer.reset().ok();
+                Err(error.into())
+            }
+        }
+    } else {
+        let program = compile_no_check(context, config, main);
+        if program.is_err() {
+            return Ok(());
+        }
```

Comment:
This should only check for a unsatisfied constraint. This allows a test to pass with invalid Noir code.

Reply:
This should be a passing test as it's valid Noir but an unsatisfiable program but 

```rust
#[test(should_fail)]
fn test() {
    let x: Field = 100; 
    assert(x == true);
}
```

should fail as it's invalid Noir.